### PR TITLE
Enumerate shapes feature in fast sygus enumerator 

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the feature of enumerating shapes in the fast sygus enumerator, which is required for a new algorithm for sygus solution reconstruction.

This also adds a builtinToSygus backwards mapping that is required for that algorithm as well.

Note this also changes the implementation of mkFreeVar in sygus term database from skolem to bound variable, which is required to do proper caching for expr::hasBoundVar.

FYI @abdoo8080 .

### CODE REVIEW

File: src/theory/quantifiers/sygus/sygus_enumerator.cpp [hunk 1]
```diff
@@ -995,6 +1025,106 @@ bool SygusEnumerator::TermEnumMaster::initializeChild(unsigned i,
...
+    std::vector<Node>& children)
+{
+  if (children.size() <= 2)
+  {
+    // don't need to convert constants and unary applications
+    return;
+  }
+  std::map<TypeNode, int> vcounter;
+  // buffered child, so that we only compute vcounter if there are more than
+  // one children with free variables, since otherwise there is no change.
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/sygus_enumerator.cpp:1038]: Can you give an example to clarify this?

File: src/theory/quantifiers/sygus/sygus_enumerator.h [hunk 2]
```diff
@@ -393,6 +415,13 @@ class SygusEnumerator : public EnumValGenerator
     bool initializeChild(unsigned i, unsigned sizeMin);
     /** increment internal, helper for increment() */
     bool incrementInternal();
+    /**
+     * Convert children so that all sygus free variables are unique. Note that
+     * the first child is a constructor operator and should be skipped.
+     */
+    void childrenToShape(std::vector<Node>& children);
+    /** Convert n into shape based on the variable counters. */
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/sygus_enumerator.h:423]: give an example

File: src/theory/quantifiers/sygus/sygus_enumerator.h [hunk 3]
```diff
@@ -40,11 +40,23 @@ class SygusPbe;
  * builtin terms (TermDb::sygusToBuiltin) can be shown to be equivalent via
  * rewriting. It enumerates terms in order of sygus term size
  * (TermDb::getSygusTermSize).
+ *
+ * It also can be configured to enumerates sygus terms with free variables,
```
[the remark this answers is not in the document]
Comment (HanielB) [reply src/theory/quantifiers/sygus/sygus_enumerator.h:44]: Thoughts on this? Not necessarily what I suggested, but I think it'd be good to clarify how z1 etc differ from x

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/sygus/sygus_enumerator.h:44]: Ah I thought I applied this, I agree to add.

**Pull request merged**